crates/rethnet_evm_napi/src/tracer/js_tracer.rs

```diff
@@ -0,0 +1,664 @@
+                                })?;
+
+                            ctx.env.create_buffer_copy(&log.data).and_then(|data| {
+                                log_object.set_named_property("data", data.into_raw())
+                            })?;
+
+                            logs.insert(log_object)?;
+                        }
+
+                        logs.coerce_to_object()
+                    })
+                    .and_then(|logs| execution_result.set_named_property("logs", logs))?;
+
+                ctx.env
+                    .create_object()
+                    .and_then(|trace| execution_result.set_named_property("trace", trace))?;
+
+                tracing_message_result.set_named_property("executionResult", execution_result)?;
+
+                let next = ctx.env.create_object()?;
+
+                let promise = ctx.callback.call(None, &[tracing_message_result, next])?;
+                let result = await_void_promise(ctx.env, promise, ctx.value.sender);
+
+                handle_error(sender, result)
+            },
+        )?;
+
+        Ok(Self {
+            before_message_fn,
+            step_fn,
+            after_message_fn,
+            pre_step: None,
+        })
+    }
+}
+
+impl<D> rethnet_evm::Inspector<D> for JsTracer
+where
+    D: rethnet_evm::Database,
+{
+    fn call(
+        &mut self,
+        data: &mut rethnet_evm::EVMData<'_, D>,
+        inputs: &mut rethnet_evm::CallInputs,
+        _is_static: bool,
+    ) -> (Return, Gas, rethnet_eth::Bytes) {
+        let (sender, receiver) = channel();
+
+        let status = self.before_message_fn.call(
+            BeforeMessageHandlerCall {
+                depth: data.journaled_state.depth,
+                to: Some(inputs.context.address),
+                data: inputs.input.clone(),
+                value: inputs.transfer.value,
+                code_address: Some(inputs.context.code_address),
+                sender,
+            },
+            ThreadsafeFunctionCallMode::Blocking,
+        );
+        assert_eq!(status, Status::Ok);
+
+        receiver
+            .recv()
+            .unwrap()
+            .expect("Failed call to BeforeMessageHandler");
+
+        (Return::Continue, Gas::new(0), Bytes::default())
+    }
+
+    fn call_end(
+        &mut self,
+        data: &mut rethnet_evm::EVMData<'_, D>,
+        _inputs: &rethnet_evm::CallInputs,
+        remaining_gas: Gas,
+        ret: Return,
+        out: Bytes,
+        _is_static: bool,
+    ) -> (Return, Gas, Bytes) {
+        let (sender, receiver) = channel();
+
+        let status = self.after_message_fn.call(
+            AfterMessageHandlerCall {
+                result: rethnet_evm::ExecutionResult {
+                    exit_reason: ret,
+                    out: rethnet_evm::TransactOut::Call(out.clone()),
+                    gas_used: if ret == Return::InvalidOpcode || ret == Return::OpcodeNotFound {
```

Comment:
I had to do this to match ethereumjs (same pattern below). What do you think, @fvictorio, should this be handled in `revm` or here?

Reply:
Ugh, good question, I don't know. It seems to me like something revm should do, but maybe there's a good reason to do it this way. It's worth opening an issue to discuss it though.

Reply:
https://github.com/bluealloy/revm/issues/335